enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,782 @@
+
+In a Create operation, if both controllers attempted the create simultaneously, it could lead to an additional host being created and then orphaned within the infrastructure provider.
+Once a controller has created the host, it is expected to populate the provider ID and use this to look up the infrastructure in future reconciles.
+The synchronisation controller would overwrite the non-authoritative provider ID, effectively orphaning the host.
+
+In an Update operation, since the infrastructure is immutable, the updates are only gathering information.
+In this case, the synchronisation controller would overwrite the information, should it differ, in the non-authoritative API.
+This could lead to a hot loop/fight between the two controllers.
+Rate limiting should be used in the controllers to prevent rapid reconciles in this fashion.
+
+In a Delete operation, both controllers would attempt to delete the same infrastructure.
+It is expected that this would not cause issue in most cases as the infrastructure provider should be able to handle multiple requests to delete the same object.
+Where Machine lifecycle hooks are in place on Machines, there should be mirrored across both APIs and should prevent either taking action until they are removed.
+Machine lifecycle hooks are expected to continue to function as expected.
+
+From this, it is imperative that we ensure Create operations are not reconciled by multiple Machine controllers to prevent the potential of leaking resources.
+Testing will be focused on the creation flow.
+
+[filtering]: #filtering-reconciles-in-capi-controllers
+
+### Drawbacks
+
+#### Carrying filtering logic in the Cluster API controllers
+
+We aim to carry as little code in forks as possible, however, we are changing the behaviour of the controllers with the reconcile filtering logic described above.
+
+This will be a fairly major patch that we need to carry.
+We need to make sure that the patch will apply cleanly in the future and that we minimise the maintenance burden of this carry as much as possible.
+
+#### Mirroring resources may be confusing for end users
+
+We want to provide an easy method for migration for users, however, having multiple representations in OpenShift of the same infrastructure, may be confusing for end users.
+
+They may expect to be able to make changes to either resource and have them reflected.
+For that to be safe we would have to make sure all writes to objects are synchronous to both objects.
+This is technically very challenging and has been dismissed as an alternative.
+
+Documentation will be provided to explain to users the expected behaviours and the expectations to allow them to migrate their infrastructure across.
+
+## Design Details
+
+### Open Questions
+
+Q. How complex will it be to have the Cluster API controllers look up Machine API resources and prevent reconciliation when not authoritative? What will the maintenance burden be for this?
+A. TODO
```

Comment:
Seems this was my question above, ideally we don't require any new code that looks up the authoritative field but rather we could do a combination of the following:

- Make use of the paused annotation for a Machine/MachineSet in Cluster API to pause reconciliation for specific objects
- Use predicates to filter out machines that we're not authoritative for
- Ensure spec.generation and status.generation match before "handoff"

Reply:
> * Make use of the paused annotation for a Machine/MachineSet in Cluster API to pause reconciliation for specific objects

Disagree on this one. I think we need to have one place in the API where the authority is defined. Having multiple places, such as `non-authoritative means paused in CAPI` will make it difficult, as we will have to sync resources across change boundaries, making it a delicate operation to transition.

To use the pause we would have to:
* Webhook to prevent the change if the authoritative API for the resource is MAPI
* When transitioning MAPI to CAPI, stop MAPI reconciling, then un-pause CAPI
* When transitioning to CAPI, pause all resources, prevent them being un-paused, then start reconciling MAPI

It's not impossible, but it's certainly more fragile than I'd like, hence preferring a single place in the API that all consumers (MAPI and CAPI) can use to determine who is authoritative.

> * Use predicates to filter out machines that we're not authoritative for

Agreed

> * Ensure spec.generation and status.generation match before "handoff"

Agreed